### PR TITLE
Raster RDD Cropping

### PR DESCRIPTION
**Pull request by jamesmcclain**
These changes add a `crop` method to the `RasterRDD` type that allows the same to be cropped by a given extent.

### CODE REVIEW

**Comment by lossyrob** [comment]
Crop methods/code can be in `geotrellis.spark.crop`

File: spark/src/main/scala/geotrellis/spark/filter/RasterRDDCropMethods.scala [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    val md = self.metadata
+    val mt = md.mapTransform
+    val rdd = self
+      .filter({ case (key, tile) =>
+        val srcExtent = mt(SpatialKey(key.col, key.row))
+        extent.interiorIntersects(srcExtent) })
+      .map({ case (key, tile) =>
+        val srcExtent = mt(SpatialKey(key.col, key.row))
+        val newTile = tile.crop(srcExtent, extent, options)
+        (key, newTile) })
```
Comment (lossyrob) [inline spark/src/main/scala/geotrellis/spark/filter/RasterRDDCropMethods.scala]: In spark, if we do something with a Key Value RDD, we can lose the partitioner of we don't explicitly state to keep it (in the case where we are potential changing keys), or if we use functions that don't change keys.

In this case, we need the key value, so I would write this as a mapPartitions with `preservePartitioning = true`

```scala
val rdd = self
  .mapPartitions({ partition =>
    partition.flatMap { case (key, tile) =>
      val srcExtent = mt(SpatialKey(key.col, key.row))
      if(extent.interiorIntersects(srcExtent)) {
        if(extent.contains(srcExtent)) {
          Some((key, tile))
        } else {
          Some((key, tile.crop(srcExtent, extent, options)))
        }
      } else {
        None
      }
    }
  }), preservesPartitioning = true)
```

File: spark/src/main/scala/geotrellis/spark/filter/RasterRDDCropMethods.scala [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import geotrellis.raster._
+import geotrellis.raster.crop.Crop.Options
+import geotrellis.spark._
+import geotrellis.util.MethodExtensions
+import geotrellis.vector.Extent
+
+import org.apache.spark.rdd._
+
+
+abstract class RasterRDDCropMethods[K <% SpatialKey] extends MethodExtensions[RasterRDD[K]] {
```
Comment (lossyrob) [inline spark/src/main/scala/geotrellis/spark/filter/RasterRDDCropMethods.scala]: `RasterRDDCropMethods[K: SpatialComponent]`

File: spark/src/main/scala/geotrellis/spark/filter/RasterRDDCropMethods.scala [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  def crop(extent: Extent, options: Options): RasterRDD[K] = {
+    val md = self.metadata
+    val mt = md.mapTransform
+    val rdd = self
+      .filter({ case (key, tile) =>
+        val srcExtent = mt(SpatialKey(key.col, key.row))
```
Comment (lossyrob) [inline spark/src/main/scala/geotrellis/spark/filter/RasterRDDCropMethods.scala]: `val srcExtent = mt(key)`

File: spark/src/main/scala/geotrellis/spark/package.scala [hunk 4]
```diff
@@ -127,6 +128,12 @@ package object spark
       }, preservesPartitioning = true)
   }
 
+  implicit class withRasterRDDSpatialKeyCropMethods(val self: RasterRDD[SpatialKey])
+      extends RasterRDDCropMethods[SpatialKey]
+
+  implicit class withRasterRDDSpaceTimeKeyCropMethods(val self: RasterRDD[SpaceTimeKey])
```
Comment (lossyrob) [inline spark/src/main/scala/geotrellis/spark/package.scala]: these implicits should be defined in an `Implicits` trait in the package (`crop`). An object named Implicits in that package extends that (companion object), and then we extent the package with the Implicits trait

File: spark/src/test/scala/geotrellis/spark/filter/RasterRDDCropMethodsSpec.scala [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+  describe("RasterRDD Crop Methods") {
+    val path = "raster-test/data/aspect.tif"
+    val gt = SingleBandGeoTiff(path)
+    val originalRaster = gt.raster.resample(500, 500)
+    val (_, rdd) = createRasterRDD(originalRaster, 5, 5, gt.crs)
+    val md = rdd.metadata
+    val overall = md.extent
+    val half = Extent(633750, 218375, 641250, 225125)
+    val small = Extent(639000.0, 217700.0, 642000.0, 220399.0)
```
Comment (lossyrob) [inline spark/src/test/scala/geotrellis/spark/filter/RasterRDDCropMethodsSpec.scala]: Should do computation to get new extents, in the unlikely event that the extent of the underlying raster would ever change.

File: spark/src/main/scala/geotrellis/spark/filter/Implicits.scala [hunk 6]
```diff
@@ -0,0 +1,10 @@
+package geotrellis.spark.filter
+
+import geotrellis.spark.SpaceTimeKey
+
+
+object Implicits extends Implicits
+
+trait Implicits {
+  implicit def SpaceTimeToSpatial(st: SpaceTimeKey) = st.spatialKey
```
Comment (lossyrob) [inline spark/src/main/scala/geotrellis/spark/filter/Implicits.scala]: No need to have this (can use `SpatialComponent`).

This is where the implicit classes would live, then root package extends this trait

**Comment by jamesmcclain** [comment]
I believe that all comments prior to this one have been addressed.

File: spark/src/main/scala/geotrellis/spark/crop/RasterRDDCropMethods.scala [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+abstract class RasterRDDCropMethods[K: SpatialComponent] extends MethodExtensions[RasterRDD[K]] {
+  def crop(extent: Extent, options: Options): RasterRDD[K] = {
+    val md = self.metadata
+    val mt = md.mapTransform
+    val rdd = self
+      .mapPartitions({ partition =>
+        partition.flatMap({ case (key, tile) =>
+          val srcExtent = mt(key)
+          if (extent.interiorIntersects(srcExtent)) {
+            val newTile = tile.crop(srcExtent, extent, options)
```
Comment (lossyrob) [inline spark/src/main/scala/geotrellis/spark/crop/RasterRDDCropMethods.scala]: This does an unnecessary crop if the srcExtent is contained fully within the extent

Comment (jamesmcclain) [inline spark/src/main/scala/geotrellis/spark/crop/RasterRDDCropMethods.scala]: Okay, corrected.

**Pull request merged**